### DISCUSSION TITLE
How to write the design matrix when stimulus onsets are not aligned with TRs?

### DISCUSSION THREAD

**Discussion by QingfangLiu** [cvnlab/GLMsingle#61] [Q&A] [answered]
Hi I'm new in using GLMsingle. In the example, every single trial is aligned with TR so the design matrix is filled with 1 and 0. In my data, the TR is 2s, and my stimulus onsets are not aligned with the data acquisition time points. I wonder if I can still use GLMsingle in this case and if so how I should write the design matrix. Thanks!

**Comment by kendrickkay** [comment, marked as answer]
Yes, you can use it, if you can prepare alternative versions of your data and/or design matrix.  See below (this will be inserted on https://glmsingle.readthedocs.io/ soon).

We hope to have some code examples of how one might do that (e.g. interpolating your fMRI time series data) soon.  Stay tuned.

<img width="977" alt="image" src="https://user-images.githubusercontent.com/2728338/167256098-42b02678-cbfa-4c34-a0f3-128188483c17.png">

**Reply by QingfangLiu** [reply]
Thanks for your quick response. I look forward to seeing it!

**Reply by kendrickkay** [reply]
You could probably just take a stab at it yourself.  Basically, you just have to resample your fMRI data to a desired temporal sampling rate (e.g. "TR") and then think about your experimental design and reformat the design matrix accordingly.  You may have to round a little.  These functions might be helpful: https://github.com/Charestlab/pyslicetime/blob/master/slicetime/tseriesinterp.py, https://github.com/cvnlab/knkutils/blob/master/timeseries/tseriesinterp.m

**Reply by QingfangLiu** [reply]
Yes! I was planning to upsample my data to 1s (from 2s)  and then use the nearest TR. This interpolation function will help a lot. Thanks again!

**Comment by jimzhang629** [comment]
Hi!

I'd like to follow up on this discussion--
This means that if my stimulus onsets are in terms of seconds and are randomly jittered (i.e., [1.34 seconds, 4.28 seconds, 7.91 seconds]), I can't simply convert their onset times to TRs through dividing by my TR duration (2 seconds) (i.e., [1.34/2 seconds, 4.28/2 seconds, 7.91/2 seconds]). Instead, I would have to round them to the nearest TR onset time (i.e., [2 seconds, 4 seconds, 8 seconds])?

Additionally, my stimuli durations are based on RT and are thus different for each trial, even within conditions. I'm not sure how to deal with this since it seems like glmsingle needs stimuli of the same condition to have the same duration?

Thank you for any advice!

**Comment by kendrickkay** [comment]
Yes - in general, you have to work in "nearest TR" times, since the input format of the design matrix is just a matrix of onsets.

Regarding your specific case, it sounds like there may be relatively small differences in durations (contingent on RT)?  If so, it may be likely fine to just assume a fixed duration (if the RT variations are small, like up +/- 500 ms).  Ultimately, the idea is that BOLD response differences can be ultimately traced to the amplitude (and not the shape).  For small-ish time differences, this is essentially true.  For very large differences (e.g. 8-s vs 4-s neural activity durations), this will start to be false, in which case one has to re-design one's approach to the GLM strategy.    These issues are discussed a bit on the GLMsingle FAQ.

**Reply by jimzhang629** [reply]
Ah, that makes sense. Thank you for the quick response!
